modules/library/metadata/src/main/java/org/geotools/factory/CategoryRegistry.java

```diff
@@ -0,0 +1,200 @@
+package org.geotools.factory;
+
+import org.geotools.util.PartiallyOrderedSet;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static org.geotools.util.Utilities.stream;
+
+/*
+ * TODO: document all
+ *
+ *  - no null checks
+ *  - no subclass relationships between categories and instances
+ */
+
+class CategoryRegistry {
+
+	private final FactoryRegistry factoryRegistry;
+	private final Map<Class<?>, InstanceRegistry<?>> categories = new HashMap<>();
+
+	public CategoryRegistry(FactoryRegistry factoryRegistry) {
+		this.factoryRegistry = factoryRegistry;
+	}
+
+	public void registerCategory(final Class<?> category) {
+		// TODO: should anything happen if the category is already registered?
+		categories.put(category, new InstanceRegistry<>(factoryRegistry, category));
```

Comment:
That detail would need to be clarified by someone who is familiar with the current behaviour, and then added as a Javadoc to this method.

``Map.put`` will overwrite any existing ``InstanceRegistry``, and return the previous value as a result, but the result is being ignored right now. ``computeIfAbsent`` attempts to avoid that behaviour by checking if the key is there before the put occurs.

If you want thread-safe behaviour you would need to switch to a ``ConcurrentMap``, as although ``Map`` supports ``computeIfAbsent``, it doesn't have a thread-safe contract.